tower-http/src/follow_redirect/policy/mod.rs

```diff
@@ -0,0 +1,279 @@
+//! Tools for customizing the behavior of a [`FollowRedirect`][super::FollowRedirect] middleware.
+
+mod clone_body_fn;
+mod filter_credentials;
+mod join;
+mod limited;
+mod redirect_fn;
+mod same_origin;
+mod select;
+
+pub use self::{
+    clone_body_fn::{clone_body_fn, CloneBodyFn},
+    filter_credentials::FilterCredentials,
+    join::{join, Join},
+    limited::Limited,
+    redirect_fn::{redirect_fn, RedirectFn},
+    same_origin::SameOrigin,
+    select::{select, Select},
+};
+
+use http::{uri::Scheme, Request, StatusCode, Uri};
+use std::fmt;
+
+/// Trait for the policy on handling redirection responses.
+///
+/// # Example
+///
+/// Detecting a cyclic redirection:
+///
+/// ```
+/// use http::{Request, Uri};
+/// use std::collections::HashSet;
+/// use tower_http::follow_redirect::policy::{Action, Attempt, Policy};
+///
+/// #[derive(Clone)]
+/// pub struct DetectCycle {
+///     uris: HashSet<Uri>,
+/// }
+///
+/// impl<B, E> Policy<B, E> for DetectCycle {
+///     fn redirect(&mut self, attempt: &Attempt<'_>) -> Action<E> {
+///         if self.uris.contains(attempt.location()) {
+///             Action::stop()
+///         } else {
+///             self.uris.insert(attempt.previous().clone());
+///             Action::follow()
+///         }
+///     }
+/// }
+/// ```
+pub trait Policy<B, E> {
+    /// Invoked when the service received a response with a redirection status code (`3xx`).
+    ///
+    /// This method returns an [`Action`] which indicates whether the service should follow
+    /// the redirection.
+    fn redirect(&mut self, attempt: &Attempt<'_>) -> Action<E>;
+
+    /// Invoked right before the service makes a request, regardless of whether it is redirected
+    /// or not.
```

Comment:
Think we should have a little note about why this method is useful:

```suggestion
    /// or not.
    ///
    /// This can for example be used to remove sensitive headers from the request
    /// or prepare the request in other ways.
    ///
    /// The default implementation returns does nothing.
```